File: frontend/src/sdk/cameras.ts
/**
 * Safelynx SDK - Cameras API
 * 
 * Manages camera configuration and streaming.
 * Matches Rust backend: CameraResponse, CameraType, CameraStatus
 */

import { httpClient, type ApiResponse } from './client'
import type { Camera, GeoLocation } from '@/types'

// Backend response types (matching Rust structs)
export interface CameraResponse {
  id: string
  name: string
  camera_type: 'builtin' | 'usb' | 'rtsp'
  device_id: string
  rtsp_url: string | null
  location: LocationResponse | null
  status: 'active' | 'inactive' | 'error'
  resolution: ResolutionResponse
  fps: number
  is_enabled: boolean
  last_frame_at: string | null
  created_at: string
}

export interface LocationResponse {
  latitude: number
  longitude: number
  name: string | null
}

export interface ResolutionResponse {
  width: number
  height: number
}

export interface CreateCameraRequest {
  name: string
  camera_type: 'builtin' | 'usb' | 'rtsp'
  device_id: string
  rtsp_url?: string
  location?: {
    latitude: number
    longitude: number
    name?: string
  }
}

export interface UpdateCameraRequest {
  name?: string
  location?: {
    latitude: number
    longitude: number
    name?: string
  }
  resolution?: {
    width: number
    height: number
  }
  fps?: number
  enabled?: boolean
}

export interface AvailableCameraResponse {
  index: number
  name: string
  description: string
}

// Transform backend response to frontend Camera type
function toCamera(response: CameraResponse): Camera {
  return {
    id: response.id,
    name: response.name,
    camera_type: response.camera_type,
    device_id: response.device_id,
    rtsp_url: response.rtsp_url,
    status: response.status,
    streaming: response.status === 'active',
    resolution: response.resolution,
    width: response.resolution.width,
    height: response.resolution.height,
    fps: response.fps,
    is_enabled: response.is_enabled,
    location: response.location as GeoLocation | null,
    last_frame_at: response.last_frame_at,
    created_at: response.created_at,
  }
}

/**
 * Cameras SDK
 */
export const camerasSdk = {
  /**
   * List all cameras
   */
  async list(): Promise<ApiResponse<Camera[]>> {
    const response = await httpClient.get<CameraResponse[]>('/cameras')
    return {
      ...response,
      data: response.data.map(toCamera),
    }
  },

  /**
   * Get a single camera by ID
   */
  async get(id: string): Promise<ApiResponse<Camera>> {
    const response = await httpClient.get<CameraResponse>(`/cameras/${id}`)
    return {
      ...response,
      data: toCamera(response.data),
    }
  },

  /**
   * Create a new camera
   */
  async create(data: CreateCameraRequest): Promise<ApiResponse<Camera>> {
    const response = await httpClient.post<CameraResponse>('/cameras', data)
    return {
      ...response,
      data: toCamera(response.data),
    }
  },

  /**
   * Update a camera
   */
  async update(id: string, data: UpdateCameraRequest): Promise<ApiResponse<Camera>> {
    const response = await httpClient.put<CameraResponse>(`/cameras/${id}`, data)
    return {
      ...response,
      data: toCamera(response.data),
    }
  },

  /**
   * Delete a camera
   */
  async delete(id: string): Promise<ApiResponse<void>> {
    return httpClient.delete(`/cameras/${id}`)
  },

  /**
   * Start camera streaming
   */
  async startStream(id: string): Promise<ApiResponse<void>> {
    return httpClient.post(`/cameras/${id}/stream/start`)
  },

  /**
   * Stop camera streaming
   */
  async stopStream(id: string): Promise<ApiResponse<void>> {
    return httpClient.post(`/cameras/${id}/stream/stop`)
  },

  /**
   * List available system cameras
   */
  async listAvailable(): Promise<ApiResponse<AvailableCameraResponse[]>> {
    return httpClient.get<AvailableCameraResponse[]>('/cameras/available')
  },
}

File: backend/src/infrastructure/server/api/cameras.rs
//! Camera API Endpoints

use std::sync::Arc;
use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::application::use_cases::{CreateCameraRequest, UpdateCameraRequest};
use crate::domain::entities::{Camera, CameraStatus, CameraType};
use crate::domain::value_objects::GeoLocation;
use crate::infrastructure::camera::list_cameras as list_system_cameras;
use crate::infrastructure::server::AppState;

#[derive(Debug, Serialize)]
pub struct CameraResponse {
    pub id: Uuid,
    pub name: String,
    pub camera_type: CameraType,
    pub device_id: String,
    pub rtsp_url: Option<String>,
    pub location: Option<LocationResponse>,
    pub status: CameraStatus,
    pub resolution: ResolutionResponse,
    pub fps: i32,
    pub is_enabled: bool,
    pub last_frame_at: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Serialize)]
pub struct LocationResponse {
    pub latitude: f64,
    pub longitude: f64,
    pub name: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ResolutionResponse {
    pub width: i32,
    pub height: i32,
}

impl From<Camera> for CameraResponse {
    fn from(c: Camera) -> Self {
        Self {
            id: c.id(),
            name: c.name().to_string(),
            camera_type: c.camera_type(),
            device_id: c.device_id().to_string(),
            rtsp_url: c.rtsp_url().map(String::from),
            location: c.location().map(|l| LocationResponse {
                latitude: l.latitude(),
                longitude: l.longitude(),
                name: l.name().map(String::from),
            }),
            status: c.status(),
            resolution: {
                let (w, h) = c.resolution();
                ResolutionResponse { width: w, height: h }
            },
            fps: c.fps(),
            is_enabled: c.is_enabled(),
            last_frame_at: c.last_frame_at().map(|t| t.to_rfc3339()),
            created_at: c.created_at().to_rfc3339(),
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct CreateCameraBody {
    pub name: String,
    pub camera_type: CameraType,
    pub device_id: String,
    pub rtsp_url: Option<String>,
    pub location: Option<CreateLocationBody>,
}

#[derive(Debug, Deserialize)]
pub struct CreateLocationBody {
    pub latitude: f64,
    pub longitude: f64,
    pub name: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateCameraBody {
    pub name: Option<String>,
    pub location: Option<CreateLocationBody>,
    pub resolution: Option<ResolutionBody>,
    pub fps: Option<i32>,
    pub enabled: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct ResolutionBody {
    pub width: i32,
    pub height: i32,
}

#[derive(Debug, Serialize)]
pub struct AvailableCameraResponse {
    pub index: u32,
    pub name: String,
    pub description: String,
}

/// GET /api/v1/cameras
pub async fn list_cameras(
    State(state): State<Arc<AppState>>,
) -> Result<Json<Vec<CameraResponse>>, StatusCode> {
    let cameras = state
        .manage_cameras
        .list_cameras()
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    let responses: Vec<CameraResponse> = cameras.into_iter().map(Into::into).collect();

    Ok(Json(responses))
}

/// POST /api/v1/cameras
pub async fn create_camera(
    State(state): State<Arc<AppState>>,
    Json(body): Json<CreateCameraBody>,
) -> Result<(StatusCode, Json<CameraResponse>), StatusCode> {
    let request = CreateCameraRequest {
        name: body.name,
        camera_type: body.camera_type,
        device_id: body.device_id,
        rtsp_url: body.rtsp_url,
        location: body.location.map(|l| {
            GeoLocation::with_metadata(l.latitude, l.longitude, None, None, l.name)
        }),
    };

    let camera = state
        .manage_cameras
        .create_camera(request)
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok((StatusCode::CREATED, Json(camera.into())))
}

/// GET /api/v1/cameras/:id
pub async fn get_camera(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<Json<CameraResponse>, StatusCode> {
    let camera = state
        .manage_cameras
        .get_camera(id)
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?
        .ok_or(StatusCode::NOT_FOUND)?;

    Ok(Json(camera.into()))
}

/// PUT /api/v1/cameras/:id
pub async fn update_camera(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
    Json(body): Json<UpdateCameraBody>,
) -> Result<Json<CameraResponse>, StatusCode> {
    let request = UpdateCameraRequest {
        name: body.name,
        location: body.location.map(|l| {
            GeoLocation::with_metadata(l.latitude, l.longitude, None, None, l.name)
        }),
        resolution: body.resolution.map(|r| (r.width, r.height)),
        fps: body.fps,
        enabled: body.enabled,
    };

    let camera = state
        .manage_cameras
        .update_camera(id, request)
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?
        .ok_or(StatusCode::NOT_FOUND)?;

    Ok(Json(camera.into()))
}

/// DELETE /api/v1/cameras/:id
pub async fn delete_camera(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, StatusCode> {
    let deleted = state
        .manage_cameras
        .delete_camera(id)
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    if deleted {
        Ok(StatusCode::NO_CONTENT)
    } else {
        Err(StatusCode::NOT_FOUND)
    }
}

/// POST /api/v1/cameras/:id/stream/start
pub async fn start_stream(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, StatusCode> {
    state
        .manage_cameras
        .set_camera_enabled(id, true)
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(StatusCode::OK)
}

/// POST /api/v1/cameras/:id/stream/stop
pub async fn stop_stream(
    State(state): State<Arc<AppState>>,
    Path(id): Path<Uuid>,
) -> Result<StatusCode, StatusCode> {
    state
        .manage_cameras
        .set_camera_enabled(id, false)
        .await
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    Ok(StatusCode::OK)
}

/// GET /api/v1/cameras/available
pub async fn list_available_cameras() -> Json<Vec<AvailableCameraResponse>> {
    let cameras = list_system_cameras();
    
    let responses: Vec<AvailableCameraResponse> = cameras
        .into_iter()
        .map(|c| AvailableCameraResponse {
            index: c.index,
            name: c.name,
            description: c.description,
        })
        .collect();

    Json(responses)
}
